lib/Runtime/Library/LazyJSONString.cpp

```diff
@@ -0,0 +1,139 @@
+    JavascriptString(type),
+    jsonContent(jsonContent),
+    gap(gap),
+    gapLength(gapLength)
+{
+    // Use SetLength to ensure length is valid
+    SetLength(length);
+}
+
+DynamicObject*
+LazyJSONString::ParseObject(_In_ JSONObject* valueList) const
+{
+    const uint elementCount = valueList->Count();
+    PropertyIndex requestedInlineSlotCapacity = static_cast<PropertyIndex>(elementCount);
+    DynamicObject* obj = this->GetLibrary()->CreateObject(true, requestedInlineSlotCapacity);
+    FOREACH_DLISTCOUNTED_ENTRY(JSONProperty, Recycler, entry, valueList)
+    {
+        Var propertyValue = Parse(&entry);
+        PropertyValueInfo info;
+        PropertyString* propertyString = PropertyString::TryFromVar(entry.propertyName);
+        if (!propertyString || !propertyString->TrySetPropertyFromCache(obj, propertyValue, this->GetScriptContext(), PropertyOperation_None, &info))
+        {
+            JavascriptOperators::SetProperty(obj, obj, entry.propertyRecord->GetPropertyId(), propertyValue, &info, this->GetScriptContext());
+        }
+    }
+    NEXT_DLISTCOUNTED_ENTRY;
+    return obj;
+}
+
+JavascriptArray*
+LazyJSONString::ParseArray(_In_ JSONArray* jsonArray) const
+{
+    const uint32 length = jsonArray->length;
+    JavascriptArray* arr = this->GetLibrary()->CreateArrayLiteral(length);
+    JSONProperty* prop = jsonArray->arr;
+    for (uint i = 0; i < length; ++i)
+    {
+        Var element = Parse(&prop[i]);
+        BOOL result = arr->SetItem(i, element, PropertyOperation_None);
+        Assert(result); // Setting item in an array we just allocated should always succeed
+    }
+    return arr;
+}
+
+Var
+LazyJSONString::Parse(_In_ JSONProperty* prop) const
+{
+    switch (prop->type)
+    {
+    case JSONContentType::Array:
+        return ParseArray(prop->arr);
+    case JSONContentType::Null:
+        return this->GetLibrary()->GetNull();
+    case JSONContentType::True:
+        return this->GetLibrary()->GetTrue();
+    case JSONContentType::False:
+        return this->GetLibrary()->GetFalse();
+    case JSONContentType::Number:
+        return prop->numericValue.value;
+    case JSONContentType::String:
+        return prop->stringValue;
+    case JSONContentType::Object:
+        return ParseObject(prop->obj);
+    default:
+        Assume(UNREACHED);
+        return nullptr;
+    }
+}
+
+Var
+LazyJSONString::Parse() const
+{
+    // If the gap is a non-space character, then parsing will be non-trivial transformation
+    for (charcount_t i = 0; i < this->gapLength; ++i)
+    {
+        switch (this->gap[i])
+        {
+        // This is not exhaustive, just a useful subset of semantics preserving characters
+        case _u(' '):
+        case _u('\t'):
+        case _u('\n'):
+            continue;
+        default:
+            return nullptr;
+        }
+    }
+    return Parse(this->jsonContent);
+}
+
+const char16*
+LazyJSONString::GetSz()
+{
+    const charcount_t allocSize = SafeSzSize();
```

Comment:
Should this also check for whether `m_pszValue` is already set from a previous call to this function, and just return that to avoid re-serializing repeatedly? Or does that already happen somehow?

Reply:
I don't think GetSz() is generally called, and GetString() does this check, but I don't think anything prevents it, so I'll add in this check.